accounts-db/src/io_uring/dir_remover.rs

```diff
@@ -0,0 +1,258 @@
+            unix::ffi::OsStringExt as _,
+        },
+        path::PathBuf,
+    },
+};
+
+/// A fast directory remover using io_uring.
+pub struct RingDirRemover {
+    ring: Ring<State, Op>,
+}
+
+impl RingDirRemover {
+    /// Creates a new remover.
+    pub fn new() -> io::Result<RingDirRemover> {
+        let ring = IoUring::builder().setup_sqpoll(1000).build(1024)?;
+        // 12 is a good round number. Empirically removing unpacked snapshots, it appears to be high
+        // enough to be fast but not too high to cause too much inode contention.
+        ring.submitter().register_iowq_max_workers(&mut [12, 0])?;
+        Ok(Self::with_ring(ring))
+    }
+
+    /// Creates a new remover, using the provided io_uring instance.
+    pub fn with_ring(ring: IoUring) -> RingDirRemover {
+        RingDirRemover {
+            ring: Ring::new(ring, State { dirs: Slab::new() }),
+        }
+    }
+
+    /// Removes a directory and all its contents.
+    pub fn remove_dir_all(&mut self, path: impl Into<PathBuf>) -> io::Result<()> {
+        self.remove(path, true)
+    }
+
+    /// Removes the contents of a directory, but not the directory itself.
+    pub fn remove_dir_contents(&mut self, path: impl Into<PathBuf>) -> io::Result<()> {
+        self.remove(path, false)
+    }
+
+    fn remove(&mut self, path: impl Into<PathBuf>, remove_root: bool) -> io::Result<()> {
+        let path = path.into();
+
+        let root_path = path.clone();
+        let mut stack = VecDeque::new();
+
+        // iterate over all directories recursively and unlink all entries
+        stack.push_back(path);
+        while let Some(dir_path) = stack.pop_front() {
+            let file = std::fs::File::open(&dir_path)?;
+            let fd = OwnedFd::from(file);
+            let dir_fd = fd.as_raw_fd();
+            let dir_key = self.ring.ctx_mut().dirs.insert(Directory::new(fd));
+
+            for entry in fs::read_dir(&dir_path)? {
+                let entry = entry?;
+                let dir = &mut self.ring.ctx_mut().dirs[dir_key];
+
+                let is_dir = entry.file_type()?.is_dir();
+                if is_dir {
+                    stack.push_back(entry.path());
+                } else {
+                    dir.unlink_started += 1;
+                    let op = UnlinkOp::new(dir_key, dir_fd, entry.file_name());
+
+                    // Safety:
+                    // The path pointer is guaranteed to be valid for the lifetime of UnlinkOp,
+                    // which gets dropped only once it completes
+                    self.ring.push(Op::Unlink(op))?
+                }
+            }
+
+            // mark the directory as having finished scanning
+            let dir = &mut self.ring.ctx_mut().dirs[dir_key];
+            dir.set_finished_scanning();
+
+            // if the directory was empty or all unlink ops have already completed, we can close the
+            // directory fd
+            if dir.scanned_and_unlinked() {
```

Comment:
Is code ever hit? 
It seems that self.unlink_complete will only increase after we call `process_complete`, which is called inside `drain()` later outside of the loop. Inside the loop, self.unlink_complete will remain 0?